google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/application/GoogleApiClientAppEngineAdminService.java

```diff
@@ -40,18 +47,62 @@
 
   private static final String APP_ENGINE_RESOURCE_WILDCARD = "-";
   private static final long CREATE_APPLICATION_POLLING_INTERVAL_MS = 1000;
+  private static final int APPLICATION_CACHE_MAX_SIZE = 10000;
+  private final long LOCATION_CACHE_REFRESH_PERIOD_MILLIS = 1000 * 60 * 60 * 24; // 1 day
+
+  // Cache of GCP application resources. This assumes that project IDs are globally unique. Null or
+  // missing applications are not cached because they cannot be reliably invalidated.
+  private final Cache<String, Application> appEngineApplicationCache = CacheBuilder.newBuilder()
+      // arbitrary size limit
+      .maximumSize(APPLICATION_CACHE_MAX_SIZE)
+      // Even though these values should never change, it won't kill us to refresh once per day.
+      .expireAfterWrite(1, TimeUnit.DAYS)
+      .build();
+
+  // Cache of all available GCP locations. This list is expected to change very infrequently.
+  private final List<Location> appEngineLocationCache = new CopyOnWriteArrayList<>();
```

Comment:
Actually, that appears to not be an option, since an up-to-date credential is required in order to re-fetch the list of locations. I think it may actually just be better to use a regular map-based guava cache, that will just have a single value. 

I think I may also pull all of the caching logic into the parent of this class, the abstract service class. I think that logically makes a bit more sense and will help keep this code more readable.